### ISSUE TITLE
ncu with --dist-tag parameter

### ISSUE THREAD

**Issue by tarazansafak** [raineorshine/npm-check-updates#381]
Imagine that you have installed a scoped dependency via dev tag as follows;

npm install @fdp_local_safak_test/component1@dev

and component1 has following dist-tags:
dev:1.1.0
latest:2.0.0 in the artifactory.

when you run, ncu, it will return to you 2.0.0. However it would be super helpful if it would get --tag parameter to specify which dist-tag to check. By this way, it would return the 1.1.0.

Any idea?
---------------------------

**Comment by raineorshine** [comment]
I'm not familiar with scoped dependencies, but that sounds like a great enhancement! I will tag this as Enhancement so it is available to anyone who wants to contribute to npm-check-updates.

**Comment by IMalyugin** [comment]
@raineorshine
Ran into this issue, that is blocking our flow. We're making feature branches over ncu, to update monorepo to branches with specific dist-tags.

After doing a thorough research, I figured out there is no possible way to hack override dist-tags into ncu, unless it is done from within the package code.

How do I see the fix:
- Adding new option "--dist-tag";
- Passing an extra option to all the methods of package-managers (yarn/npm): newest, greatest, minor (probably except latest);
- Modifying viewMany to versions by dist tag if it exists;

I could assemble a pull request, but I might require some guidance on the testing aspect. Aka which tests do you think would be neccessary for this?

**Comment by raineorshine** [comment]
Yes, I think this is very doable. 

I believe this feature should be implemented as `--target @dev`. The `target` option determines the desired version to upgrade to. It would just involve parsing the value and the adding a `distTag` method to npm that did the same thing as `latest` but using the specified tag. 

If this is correct, the tests can be modeled after the existing `target` tests.

**Comment by IMalyugin** [comment]
I thought about that, but if we make it --target @[dist-tag] - we'll lose minor/major etc capabilities, when using dist-tag.

**Comment by raineorshine** [comment]
`--target minor` upgrades to the the highest version possible without changing the current major version. `--target @` would upgrade to the version specified by the tag. It's not clear to me what it would even mean to combine them. Upgrade to the highest version possible without surpassing the specified tag? That’s a whole new semantics.

**Comment by raineorshine** [comment]
Note that `minor/major` is already completely independent of `latest` which is technically a tag. `--target @latest` is presumed to be equivalent to `--target latest`, the default.

**Comment by IMalyugin** [comment]
> `--target minor` upgrades to the the highest version possible without changing the current major version. `--target @` would upgrade to the version specified by the tag. It's not clear to me what it would even mean to combine them. Upgrade to the highest version possible without surpassing the specified tag? That’s a whole new semantics.

Some use dist-tags more widely, aka imitating "dev", "test", "rc" branches. Generalizing, it is possible that a whole flow would work around specific dist-tags.

Latest is special, there may only be one latest revision.

On the other side, most of the settings with dist-tags only involve latest version within dist-tag, so we could keep it simple, till there is a wider need.

Having --target @dist-tag surely makes the PR multiple times easier. Just have to copy latest as "scope", and then making latest call scope('latest').

**Comment by raineorshine** [comment]
> Some use dist-tags more widely, aka simulating "dev", "test", "rc" branches. Generalizing, it is possible that a whole flow would work around specific dist-tags.

Yes, makes sense. 

> Latest is special, there may only be one latest revision.

`latest` points to a specific version and changes over time like all `dist-tags`, no?

> Having --target @[dist-tag] surely makes the PR multiple times easier. Just have to copy latest as "distTag", and then have latest call distTag('latest')

I think this is a good approach to take. To allow for more complex target specifications, we could have more information passed to the user-specified `target` function. I generally wait until there is sufficient user demand or a self-motivated contributor before adding much additional functionality.

**Comment by IMalyugin** [comment]
Agreed. Okay, I'll see what I can do, hopefully will roll out a draft PR today.

Thank you
